File: src/game/party.ts
import type { PlayerSession } from './player.js';

export class PartyManager {
  private leaderOf = new Map<string, Set<string>>();
  private memberOf = new Map<string, string>();
  private invites = new Map<string, string>();

  invite(leader: PlayerSession, target: PlayerSession): string | null {
    if (this.memberOf.has(target.username.toLowerCase())) {
      return `${target.username} is already in a party.`;
    }
    const leaderKey = leader.username.toLowerCase();
    const memberLeader = this.memberOf.get(leaderKey);
    if (memberLeader && memberLeader !== leaderKey) {
      return 'Only the party leader can invite.';
    }
    this.ensureParty(leader.username);
    this.invites.set(target.username.toLowerCase(), leaderKey);
    return null;
  }

  join(player: PlayerSession, leaderName: string): string | null {
    const leaderKey = leaderName.toLowerCase();
    if (this.invites.get(player.username.toLowerCase()) !== leaderKey) {
      return `No party invite from ${leaderName}.`;
    }
    this.invites.delete(player.username.toLowerCase());
    this.ensureParty(leaderName);
    this.leaderOf.get(leaderKey)!.add(player.username.toLowerCase());
    this.memberOf.set(player.username.toLowerCase(), leaderKey);
    return null;
  }

  leave(player: PlayerSession): void {
    const leaderKey = this.memberOf.get(player.username.toLowerCase());
    if (!leaderKey) return;
    const party = this.leaderOf.get(leaderKey);
    party?.delete(player.username.toLowerCase());
    this.memberOf.delete(player.username.toLowerCase());
    if (leaderKey === player.username.toLowerCase()) {
      const members = [...(party ?? [])];
      if (members.length > 0) {
        const newLeader = members[0];
        const rest = new Set(members.slice(1));
        this.leaderOf.delete(leaderKey);
        this.leaderOf.set(newLeader, rest);
        for (const m of members) this.memberOf.set(m, newLeader);
      } else {
        this.leaderOf.delete(leaderKey);
      }
    }
  }

  getLeaderKey(player: PlayerSession): string | null {
    return this.memberOf.get(player.username.toLowerCase()) ?? null;
  }

  getMemberUsernames(leaderKey: string, players: Map<string, PlayerSession>): string[] {
    const party = this.leaderOf.get(leaderKey);
    const leader = players.get(leaderKey);
    if (!party || !leader) return leader ? [leader.username] : [];
    const names = [leader.username];
    for (const m of party) {
      const p = players.get(m);
      if (p) names.push(p.username);
    }
    return names;
  }

  getPartyPeers(player: PlayerSession, players: Map<string, PlayerSession>): PlayerSession[] {
    const leaderKey = this.memberOf.get(player.username.toLowerCase()) ?? player.username.toLowerCase();
    return this.getMemberUsernames(leaderKey, players)
      .map((n) => players.get(n.toLowerCase()))
      .filter((p): p is PlayerSession => !!p && p.authenticated);
  }

  private ensureParty(leaderUsername: string): void {
    const key = leaderUsername.toLowerCase();
    if (!this.leaderOf.has(key)) {
      this.leaderOf.set(key, new Set());
      this.memberOf.set(key, key);
    }
  }

  onDisconnect(player: PlayerSession): void {
    this.leave(player);
    this.invites.delete(player.username.toLowerCase());
    for (const [target, leader] of this.invites) {
      if (leader === player.username.toLowerCase()) this.invites.delete(target);
    }
  }
}

File: crates/realm-core/src/party.rs
use std::collections::{HashMap, HashSet};

use crate::player::PlayerSession;

pub struct PartyManager {
    leader_of: HashMap<String, HashSet<String>>,
    member_of: HashMap<String, String>,
    invites: HashMap<String, String>,
}

impl Default for PartyManager {
    fn default() -> Self {
        Self::new()
    }
}

impl PartyManager {
    pub fn new() -> Self {
        Self {
            leader_of: HashMap::new(),
            member_of: HashMap::new(),
            invites: HashMap::new(),
        }
    }

    pub fn invite(&mut self, leader: &PlayerSession, target: &PlayerSession) -> Option<String> {
        let target_key = target.username().to_lowercase();
        if self.member_of.contains_key(&target_key) {
            return Some(format!("{} is already in a party.", target.username()));
        }

        let leader_key = leader.username().to_lowercase();
        if let Some(member_leader) = self.member_of.get(&leader_key) {
            if member_leader != &leader_key {
                return Some("Only the party leader can invite.".into());
            }
        }

        self.ensure_party(leader.username());
        self.invites.insert(target_key, leader_key);
        None
    }

    pub fn join(&mut self, player: &PlayerSession, leader_name: &str) -> Option<String> {
        let leader_key = leader_name.to_lowercase();
        let player_key = player.username().to_lowercase();

        if self.invites.get(&player_key) != Some(&leader_key) {
            return Some(format!("No party invite from {leader_name}."));
        }

        self.invites.remove(&player_key);
        self.ensure_party(leader_name);
        self.leader_of
            .get_mut(&leader_key)
            .expect("party exists after ensure_party")
            .insert(player_key.clone());
        self.member_of.insert(player_key, leader_key);
        None
    }

    pub fn leave(&mut self, player: &PlayerSession) {
        let player_key = player.username().to_lowercase();
        let Some(leader_key) = self.member_of.get(&player_key).cloned() else {
            return;
        };

        if let Some(party) = self.leader_of.get_mut(&leader_key) {
            party.remove(&player_key);
        }
        self.member_of.remove(&player_key);

        if leader_key == player_key {
            let members: Vec<String> = self
                .leader_of
                .get(&leader_key)
                .map(|set| set.iter().cloned().collect())
                .unwrap_or_default();

            if let Some(new_leader) = members.first().cloned() {
                let rest: HashSet<String> = members.iter().skip(1).cloned().collect();
                self.leader_of.remove(&leader_key);
                self.leader_of.insert(new_leader.clone(), rest);
                for member in &members {
                    self.member_of
                        .insert(member.clone(), new_leader.clone());
                }
            } else {
                self.leader_of.remove(&leader_key);
            }
        }
    }

    pub fn get_leader_key(&self, player: &PlayerSession) -> Option<String> {
        self.member_of.get(&player.username().to_lowercase()).cloned()
    }

    pub fn get_member_usernames(
        &self,
        leader_key: &str,
        players: &HashMap<String, PlayerSession>,
    ) -> Vec<String> {
        let leader_key = leader_key.to_lowercase();
        let Some(party) = self.leader_of.get(&leader_key) else {
            return players
                .get(&leader_key)
                .map(|p| vec![p.username().to_string()])
                .unwrap_or_default();
        };

        let Some(leader) = players.get(&leader_key) else {
            return Vec::new();
        };

        let mut names = vec![leader.username().to_string()];
        for member_key in party {
            if let Some(p) = players.get(member_key) {
                names.push(p.username().to_string());
            }
        }
        names
    }

    pub fn get_party_peers(
        &self,
        player: &PlayerSession,
        players: &HashMap<String, PlayerSession>,
    ) -> Vec<PlayerSession> {
        let leader_key = self
            .member_of
            .get(&player.username().to_lowercase())
            .cloned()
            .unwrap_or_else(|| player.username().to_lowercase());

        self.get_member_usernames(&leader_key, players)
            .into_iter()
            .filter_map(|name| players.get(&name.to_lowercase()).cloned())
            .filter(|p| p.authenticated)
            .collect()
    }

    fn ensure_party(&mut self, leader_username: &str) {
        let key = leader_username.to_lowercase();
        if !self.leader_of.contains_key(&key) {
            self.leader_of.insert(key.clone(), HashSet::new());
            self.member_of.insert(key.clone(), key);
        }
    }

    pub fn on_disconnect(&mut self, player: &PlayerSession) {
        let player_key = player.username().to_lowercase();
        self.leave(player);
        self.invites.remove(&player_key);

        let stale: Vec<String> = self
            .invites
            .iter()
            .filter_map(|(target, leader)| {
                if leader == &player_key {
                    Some(target.clone())
                } else {
                    None
                }
            })
            .collect();

        for target in stale {
            self.invites.remove(&target);
        }
    }
}